src/lib.rs

```diff
@@ -0,0 +1,532 @@
+/// combining all the deltas together. This is called "collapsing".
+///
+/// This function confirms that two state groups mappings lead to the
+/// exact same entries for each state group after collapsing them down.
+///
+/// # Arguments
+/// * `old_map` -   The state group data currently in the database
+/// * `new_map` -   The state group data that the old_map is being compared
+///                 to
+fn check_that_maps_match(
+    old_map: &BTreeMap<i64, StateGroupEntry>,
+    new_map: &BTreeMap<i64, StateGroupEntry>,
+) {
+    println!("Checking that state maps match...");
+
+    let pb = ProgressBar::new(old_map.len() as u64);
+    pb.set_style(
+        ProgressStyle::default_bar().template("[{elapsed_precise}] {bar} {pos}/{len} {msg}"),
+    );
+    pb.set_message("state groups");
+    pb.enable_steady_tick(100);
+
+    // Now let's iterate through and assert that the state for each group
+    // matches between the two versions.
+    old_map
+        .par_iter() // This uses rayon to run the checks in parallel
+        .try_for_each(|(sg, _)| {
+            let expected = collapse_state_maps(&old_map, *sg);
+            let actual = collapse_state_maps(&new_map, *sg);
+
+            pb.inc(1);
+
+            if expected != actual {
+                println!("State Group: {}", sg);
+                println!("Expected: {:#?}", expected);
+                println!("actual: {:#?}", actual);
+                Err(format!("State for group {} do not match", sg))
+            } else {
+                Ok(())
+            }
+        })
+        .expect("expected state to match");
+
+    pb.finish();
+
+    println!("New state map matches old one");
+}
+
+/// Gets the full state for a given group from the map (of deltas)
+fn collapse_state_maps(map: &BTreeMap<i64, StateGroupEntry>, state_group: i64) -> StateMap<Atom> {
+    let mut entry = &map[&state_group];
+    let mut state_map = StateMap::new();
+
+    let mut stack = vec![state_group];
+
+    while let Some(prev_state_group) = entry.prev_state_group {
+        stack.push(prev_state_group);
+        if !map.contains_key(&prev_state_group) {
+            panic!("Missing {}", prev_state_group);
+        }
+        entry = &map[&prev_state_group];
+    }
+
+    for sg in stack.iter().rev() {
+        state_map.extend(
+            map[&sg]
+                .state_map
+                .iter()
+                .map(|((t, s), e)| ((t, s), e.clone())),
+        );
+    }
+
+    state_map
+}
+
+// PyO3 INTERFACE STARTS HERE
+
+impl Config {
+    /// Converts string and bool arguments into a Config struct
+    ///
+    /// This function panics if db_url or room_id are empty strings!
+    pub fn new(
+        db_url: String,
+        room_id: String,
+        output_file: Option<String>,
+        max_state_group: Option<i64>,
+        min_saved_rows: Option<i32>,
+        transactions: bool,
+        level_sizes: String,
+    ) -> Result<Config, String> {
```

Comment:
We'll probably want to start using a proper error type at some point, e.g. using `anyhow` lib, but this is fine for now